src/distdl/nn/mixins/halo_mixin_new.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+
+from distdl.utilities.slicing import compute_start_index
+from distdl.utilities.slicing import compute_subshape
+
+
+class HaloMixin:
+
+    def _compute_exchange_info(self,
+                               halo_shape,
+                               neighbor_halo_shapes):
+
+        recv_buffer_shape = halo_shape.copy()
+        send_buffer_shape = np.zeros_like(halo_shape, dtype=int)
+
+        dims = len(halo_shape)
+        for dim in range(dims):
+            left_halo_shape, right_halo_shape = neighbor_halo_shapes[dim]
+
+            if left_halo_shape is not None:
+                send_buffer_shape[dim, 0] = left_halo_shape[dim, 1]
+
+            if right_halo_shape is not None:
+                send_buffer_shape[dim, 1] = right_halo_shape[dim, 0]
+
+        return recv_buffer_shape, send_buffer_shape
+
+    def _left_pad_to_length(self, array, length, value):
+        lpad = length - len(array)
+        return np.pad(array,
+                      pad_width=(lpad, 0),
+                      mode='constant',
+                      constant_values=value)
+
+    def _compute_local_start_stop_indices(self, subtensor_shapes_unbalanced, x_local_shape):
+        # this could use a refactor at some point
+        x_local_start_index = np.zeros_like(x_local_shape, dtype=np.int)
+        dims = len(x_local_shape)
+        for dim in range(2, dims):
+            for i in range(self.P_x.index[dim]):
+                idx = tuple(i if j == dim else 0 for j in range(dims))
+                x_local_start_index[dim] += subtensor_shapes_unbalanced[idx][dim]
+        x_local_stop_index = x_local_start_index + x_local_shape - 1
+        return x_local_start_index, x_local_stop_index
+
+    def _compute_needed_ranges(self, tensor_shape, halo_shape):
+
+        ranges = np.zeros_like(halo_shape)
+
+        # If we have a negative halo on the left, we want to not pass that
+        # data to the torch layer
+        ranges[:, 0] = -1*np.minimum(0, halo_shape[:, 0])
+
+        # The stop of the slice will be the data + the length of the two halos
+        # and the last maximum is so that we dont shorten the stop (keeps the
+        # parallel and sequential behavior exactly the same, but I dont think
+        # it is strictly necessary)
+        ranges[:, 1] = tensor_shape[:] + np.maximum(0, halo_shape[:, 0]) + np.maximum(0, halo_shape[:, 1])
```

Comment:
See the upsample code, this line does not allow negative right halos as written.  That code does.

Reply:
This was taken straight from HaloMixin, but I agree, it doesn't account for negative right halo correctly. However, the layer will still produce a correct result. If there is a negative right halo, then consider the last kernel placement in that partition. The kernel placement that comes after that one must extend outside of the input with right halo padding (which should be 0), which means the PyTorch conv layer will not be able to compute that kernel. Then, the right neighboring worker will take on that computation. So, in the case of negative right halos, a right halo of 0 can be set and the layer will be correct.

Still, I think it's worth looking at this in the future, so I'll add a comment.